### DISCUSSION TITLE
Loading XGBClassifier saved in the artifactstore

### DISCUSSION THREAD

**Discussion by jaysonpryde** [mlflow/mlflow#5754] [Q&A] [answered]
Given that my use case/usage to the dumped XGBClassifier model by mlflow is different, I am not planning to use:

1. **mlflow models serve**, nor
2. **mlflow.pyfunc.load_model**

to expose/consume the model. So I need this to load in the more usual way.
What I tried via python shell/ipython is loading it via:

- **pickle**
```
import  pickle
from pathlib import Path
import joblib

model = None
model_path = "/tmp/data_1/artifacts/1/729dd9ee00d741b79fa00ad01939eb4b/artifacts/model/model.xgb"
with Path(model_path).open("rb") as f:
    model = pickle.load(f)
```
- **joblib**
```
import joblib

model = joblib.load(model_path)
```
However, for both, I am just getting exceptions.
Given this, how can I load the saved model then?

Thanks in advance!

**Comment by jaysonpryde** [comment]
I tried loading a **DecisionTreeClassifier** and **RandomForestClassifier** using both **pickle** and **joblib**, and they worked just fine. So I guess this is specific to XGBClassifier, then? or just **mlflow.xgboost.autolog()**... I'll retry XGBClassifier and just use **mlflow.autolog()**.. Will give an update once done

**Reply by jaysonpryde** [reply]
According to [this](https://www.mlflow.org/docs/latest/python_api/mlflow.xgboost.html), current xgboost support is **0.90 <= xgboost <= 1.5.2**. This perhaps may be the issue since I am currently using 1.6.0... will double check this

### UPDATE
I already tested for **xgb version 1.4.0**. Unfortunately, exceptions still occur and the XGBClassifier model is not loaded.
So really need some help / guidance here, how XGB models can be loaded

**Comment by BenWilson2** [comment, marked as answer]
Does `mlflow.xgboost.load_model(model_uri=model_path)` not work for you?

**Reply by jaysonpryde** [reply]
Hi @BenWilson2 , thanks a lot for the feedback. The `load_model` is working fine. However and as mentioned, I have a different use case, where in I need to load the model using `joblib`... joblib can load the RandomForest and DecisionTree classifiers. It's just the XGBClassifier.
Thanks again

**Reply by BenWilson2** [reply]
I can't imagine that it's the version conflict (we're running nightly integration tests against XGBoost==1.6.1 and they've been passing just fine - although we don't run anything through joblib). 
The version definition within the XGBoost loader might have some issues with the flavor-defined logged version. Could you paste the stack trace that you're getting here so that we can try to figure out what's going on? 
Alternatively to get this setup to work, you might want to try serializing the model outside of the `log_artifacts()` command to support your use case in the meantime. 
I'll try to take a look at the stack trace once you post it. Thanks!

**Reply by jaysonpryde** [reply]
Thanks. Your recommendation of serializing/dumping it outside `log_artifacts()` is actually what I did as a workaround prior your reply.
But still, I'm still curious of why this happens for XGBClassifier only.

Here are lib details, model details, and the stack traces for **joblib**, **pickle** and **cloudpickle*:

### Lib Details

- **joblib** == 1.0.1
- **cloudpickle** -== 2.0.0
- **xgboost** == 1.4.0

### Model details

- **model filesize** == 24MB
- **file format** --> i don't know what specific filetype is this, but it starts with "binf" (binf^@^@^@?7^@^@^@^@^@^@^@^A^@^@^@^@^@^@) when opened with an editor. So I guess binary format???

### Joblib stacktrace
```
Traceback (most recent call last):
  File "<ipython-input-13-a26362137464>", line 2, in <module>
    clf =  joblib.load("./model.xgb")
  File "lib/python3.8/site-packages/joblib/numpy_pickle.py", line 585, in load
    obj = _unpickle(fobj, filename, mmap_mode)
  File "lib/python3.8/site-packages/joblib/numpy_pickle.py", line 504, in _unpickle
    obj = unpickler.load()
  File "lib/python3.8/pickle.py", line 1210, in load
    dispatch[key[0]](self)
  File "lib/python3.8/site-packages/joblib/numpy_pickle.py", line 329, in load_build
    Unpickler.load_build(self)
  File "lib/python3.8/pickle.py", line 1699, in load_build
    state = stack.pop()
IndexError: pop from empty list
```

### pickle stacktrace
```
Traceback (most recent call last):
  File "<ipython-input-14-0810334fc0bc>", line 3, in <module>
    clf = pickle.load(f)
_pickle.UnpicklingError: unpickling stack underflow
```

### cloudpickle stacktrace
```
Traceback (most recent call last):
  File "<ipython-input-15-651afbbd48e6>", line 3, in <module>
    clf = cloudpickle.load(f)
_pickle.UnpicklingError: unpickling stack underflow
```

**Reply by BenWilson2** [reply]
XGBoost doesn't use pickling anymore (they abandoned that in 1.0.0 release). The mechanism in use for saving models within MLflow is the sklearn wrapper interface for XGBoost which uses a custom JSON-based serialization approach which is something that all three approaches you're trying just can't handle.

The mechanism from within xgboost: 

https://github.com/dmlc/xgboost/blob/142a208a90357758216b436686a6dd0f77409eec/python-package/xgboost/sklearn.py#L730-L756

Line 747 from that snippet shows the `json.dumps(<object dict>)` serialization going on. 

I would recommend using the official loaders (we wrap the loader in MLflow but you can use the native `load_model` from within `xgboost` to deserialize your model instance that's been logged.

**Reply by jaysonpryde** [reply]
I tried loading with the `load_model` and it worked fine. Thanks a lot @BenWilson2 for the replies and guidance.
Greatly appreciate them
